Transaction feature enhancements

aaron-garnett opened the pull request:
- ~~Revised an 'if' statement filtering the Transactions class iterator that was preventing shares-only transactions (such as 'Add Shares' and 'Remove Shares') from being returned~~ (this revision failed testing, was proven to be unnecessary, and then reverted by the second commit as noted in another comment below)
- Added additional fields to Transactions to improve investment related reporting: typeKey, commission, and costbasis
- Added numerator and denominator fields to Transactions which store the ratio used for stock splits
- Added additional fields to Transactions to provide additional information about cash transactions: splitNote, transferAcctName, transferAcctKey, and tags.
- New Transfers and UserTags classes were added to facilitate these features (these classes are only called when the Transaction class is called).

Note: I am extremely grateful for the work of @HarryDolan and other contributors on this project and I would be delighted if my contributions are useful for others.

HarryDolan commented:
Wow!  Someone is actually using my code?  I'm flattered.  :-)

As you can see it's been a long time since I've done any significant work on it.  But I will review your pull request.

Thank you for your contribution.

Harry

tim-rohrer commented:
I wanted to use it back in 2022, but didn't get any response to my question.

I do believe this work has great potential. Quicken's approach to tables is convoluted and  your work really could be powerful for those who want to move away from Quicken or use the data in other apps. That is what I was working on.

@aaron-garnett, I'd love to hear what you're trying to do?

aaron-garnett commented:
@tim-rohrer My ultimate goal is to perform reporting and analysis of my income, spending, businesses, and investments in ways that Quicken cannot do on its own. I've started doing this using data visualization tools such as Power BI, by exporting data from the Quicken UI. Qquery enables me to extract the data directly from the Quicken database. In order to create some additional reporting features, such as the "Invest Amt" column that is visible in Quicken, in my own code, I forked this repo to expose the numerator, denominator, type, and costbasis fields that are attached to investment transactions. I have experience with relational databases and the Quicken database is complex but not impossible to reverse engineer.

HarryDolan commented:
Sorry, Tim, I only just yesterday noticed your question.  I had not been paying much attention to this project back then.

HarryDolan commented:
Does anyone have any Quicken test data they would be willing to share?

@aaron-garnett, it would be especially useful to have a Quicken file that contains transactions of the type your new functionality handles.

aaron-garnett commented:
@HarryDolan I sent some sample data to your email via WeTransfer. Please confirm receipt.

HarryDolan commented:
I just got it.  Thank you!

I will continue work on the project this afternoon.

Harry


> On Feb 26, 2025, at 8:16 AM, Aaron Garnett ***@***.***> wrote:
> 
> 
> aaron-garnett
>  left a comment 
> (HarryDolan/qquery#4)
> @HarryDolan I sent some sample data to your email via WeTransfer. Please confirm receipt.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.
> 
>  <https://github.com/HarryDolan> <https://github.com/HarryDolan/qquery/pull/4#issuecomment-2685155075> <https://github.com/notifications/unsubscribe-auth/AEEWH2EM3T7CUKREO43ZJUL2RXEDPAVCNFSM6AAAAABXWZVQIGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBVGE2TKMBXGU>
> 
> aaron-garnett
>  left a comment 
> (HarryDolan/qquery#4)
>  <https://github.com/HarryDolan/qquery/pull/4#issuecomment-2685155075>
> @HarryDolan <https://github.com/HarryDolan> I sent some sample data to your email via WeTransfer. Please confirm receipt.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/HarryDolan/qquery/pull/4#issuecomment-2685155075>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEEWH2EM3T7CUKREO43ZJUL2RXEDPAVCNFSM6AAAAABXWZVQIGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBVGE2TKMBXGU>.
> You are receiving this because you were mentioned.
>

HarryDolan commented:
I tried your changes with your data.  I used the command_line.py with the --list-transactions option.
It failed on the second transaction, I think on line 463.  It seems that trans['splitCategoryKey'] is Null.

aaron-garnett commented:
Ok I will perform my own testing this weekend, specifically with CLI which I have not yet used - only using Python class/method calls so far. Should be solvable.

aaron-garnett commented:
I have not been able to replicate the error. I ran the following two commands successfully from my Mac terminal:
`qquery --qdb="/Users/Shared/temp/Quicken Test Data.quicken/data" --list-accounts`
`qquery --qdb="/Users/Shared/temp/Quicken Test Data.quicken/data" --list-transactions`
I've sent the results of the above to you for analysis.

I've also noted that there are 47 transactions in the test dataset with a null value for `trans['splitCategoryKey']`, all of which are Quicken 'placeholder' transactions which are not visible in the UI. These also have null values for `trans['splitAmount']` and `trans['parentSecurityShares']` so they should be ignored by the class iterator due to line 454 in __init__.py:
            `if trans['splitAmount']==None and trans['parentSecurityShares']==None: continue`

aaron-garnett commented:
With further testing I was able to replicate the error and determine its cause. Some of my misleading test results were confounded by file versioning issues which I have since corrected. I have reverted line 454 to the logic as originally written after determining that the additional condition I added did not have the intended effect. I have posted another commit to effect this change.

aaron-garnett commented:
My second commit resolves the errors originally reported by @HarryDolan. In short, there is an exception case where a parent transaction has no split transactions thus has no transaction amount. The code as originally written would exclude these from the `_Transactions` class iterator, but my first commit on this branch added an additional condition which negated that exclusion logic and let invalid transactions flow through and trigger a Key error. The second commit reverts to the original logic. Shares only transactions still come through as valid.

HarryDolan commented:
@aaron-garnett: thank you for your contribution and all the work that went in to it.  I have merged your pull request as submitted.

aaron-garnett commented:
@HarryDolan Thank you for your prompt review and merge, and for your clean and easy to read code. It was a pleasure to contribute to this project.
